Trees that grow notes (#1419)
tweag/nickel: notes/trees-that-grow-proc-macro.md

> As nickel grows, it becomes unwieldy to use a single AST for every stage of compilation. Some things are kept out of the AST, even though they would be useful in some places, because they don't make sense and would get in the way at others (e.g. types for the LSP, closures at runtime). Other things are in the AST, but aren't actually filled in until later on in the process (e.g. forall's var_kind). This causes `unwrap()`s and `unreachable!()`s and hurts code maintainability.

Reviewer: Nitpick, but Nickel is (not yet?) compiled :stuck_out_tongue: 

```suggestion
As Nickel grows, it becomes unwieldy to use a single abstract syntax tree (AST) for every stage of the pipeline. Some data are kept out of the AST, even though they would be useful in some places, because they don't make sense and would get in the way at others (e.g. types inferred during typechecking for the LSP, a closure - i.e. a term together with its environment - at runtime). Other data placeholder are in the AST, but aren't actually filled in until later on in the process (e.g. forall's var_kind). This causes `unwrap()`s and `unreachable!()`s and hurts code maintainability.
```

## Strategies

> One strategy to deal with this is to repeat a large AST enum in multiple places with small differences, and convert from one to another as we move through compilation. This creates a large maintenance burden, since many changes have to be propagated to all the trees.

Reviewer: ```suggestion
One strategy to deal with this is to repeat a large AST enum in multiple places with small differences, and convert from one to another as we move through program manipulations. This creates a large maintenance burden, since many changes have to be propagated to all the trees.
```

> Another strategy is outlined by the paper [Trees that Grow](https://www.microsoft.com/en-us/research/uploads/prod/2016/11/trees-that-grow.pdf). This describes an extendable tree structure using type families and pattern synonyms for ergonomics. While this is quite a bit DRYer, there is still a decent amount of boilerplate, and actually using these trees is less intuitive. Especially in rust, where there are no pattern synonyms.

Reviewer: ```suggestion
Another strategy is outlined by the paper [Trees that Grow](https://www.microsoft.com/en-us/research/uploads/prod/2016/11/trees-that-grow.pdf). This describes an extendable tree structure using type families and pattern synonyms for ergonomics. While this is quite a bit DRYer, there is still a decent amount of boilerplate, and actually using these trees is less intuitive. Especially in Rust, where there are no pattern synonyms (the paper is targeting Haskell).
```

I (@Radvendii) have explored using proc macros to get the best of both worlds, and remove the boilerplate along the way. There are three ways we could use proc macros, but they share a common syntax.

## Macro Syntax

The naming of things is up in the air, but the general idea is to annotate the tree definition with how fields get modified in the extensions. There are two kinds of extensions, product extensions which add data to an existing term type, and sum extensions which add a new term type.

> ```rust

Reviewer: It could help to add a written description of what the following program is supposed to mean, that is:

> The following example models 3 different ASTs, such that ...

It's not obvious what `Foo` is here (a type? a name? something else), and it's a bit hard to follow the translation if we don't really know what we are translating to begin with.

> As Nickel grows, it becomes unwieldy to use a single abstract syntax tree (AST) for every stage of the pipeline. Some data are kept out of the AST, even though they would be useful in some places, because they don't make sense and would get in the way at others (e.g. types inferred during typechecking for the LSP, a closure - i.e. a term together with its environment - at runtime). Other data placeholder are in the AST, but aren't actually filled in until later on in the process (e.g. forall's var_kind). This causes `unwrap()`s and `unreachable!()`s and hurts code maintainability.

Reviewer 2: ```suggestion
As Nickel grows, it becomes unwieldy to use a single abstract syntax tree (AST) for every stage of the pipeline. Some data are kept out of the AST, even though they would be useful in some places, because they don't make sense and would get in the way at others (e.g. types inferred during typechecking for the LSP, a closure - i.e. a term together with its environment - at runtime). Other data placeholders are in the AST, but aren't actually filled in until later on in the process (e.g. forall's var_kind). This causes `unwrap()`s and `unreachable!()`s and hurts code maintainability.
```

```rust
enum ExpBase {
    Lit(i32),
    Var(Var),
    Ann(Box<ExpBase>, Typ),
    Abs(Var, Box<ExpBase>),
    App(Box<ExpBase, Box<ExpBase>),
}

enum ExpTC {
    Lit(i32),
    Var(Var),
    Ann(Box<ExpTC>, Typ),
    Abs(Var, Box<ExpTC>),
    App(Typ, Box<ExpTC, Box<ExpTC>),
}

enum ExpPE {
    Lit(i32),
    Var(Var),
    Ann(Box<ExpPE>, Typ),
    Abs(Var, Box<ExpPE>),
    App(Box<ExpPE, Box<ExpPE>),
    Evl(Val),
}
```

This can be used as follows

```rust
fn foo(term: ExpPE) {
    match term {
        ExpPE::Lit(l) => ...,
        ExpPE::Var(v) => ...,
        ExpPE::Ann(e, t) => ...,
        ExpPE::Abs(v, e) => ...,
        ExpPE::App(f, a) => ...,
        ExpPE::Evl(v) => ...,
    }
}
```

But note that if we want to implement similar functions for all of the tree types, we have to implement them for each type separately.

Alternatively, we could generate trees-that-grow-style trees.

```rust
// uninhabited type. cannot be constructed.
pub enum Void {}
trait Extension {
    type Lit;
    type Var;
    type Ann;
    type Abs;
    type App;
    type More;
}
enum Exp<Ext: Extension> {
    Lit(Ext::Lit, i32),
    Var(Ext::Var, Var),

>     Ann(Ext::Ann, Box<Exp>, Typ),

Reviewer 2: ```suggestion
    Ann(Ext::Ann, Box<Exp<Ext>>, Typ),
```
(and also below)
